File: external/Gondan-BRM-2009/rmiperm.r
#
# Permutation test for the race model inequality
#
# Tested with R-2.9.1, download at http://www.stats.bris.ac.uk/R/
#

# tmax statistic

tmax = function(di)
{
    # t-Tests for each percentile
    tvalues = rowMeans(di) / apply(di, 1, sd) * sqrt(ncol(di))
    max(tvalues)
} # tmax

# The significance test
#
# - d: list including data of all participants
# - quantiles: quantiles of AV distribution at which the RMI is to be tested
# - nperm: number of random permutations used to generate H0 distribution
# - return value: tmax, critical value for one-sided test, P value

rmi_perm = function(d, quantiles=c(.05, .10, .15, .20, .25, .30), nperm=10001)
{
    # Determine quantiles from entire response time distribution 
    tmix     = lapply(d, unlist) 
    q        = lapply(tmix, quantile, probs=quantiles, type=4)

    # Evaluate distribution functions at q
    FA       = mapply(function(ti, qi) ecdf(ti$A)(qi), d, q)
    FV       = mapply(function(ti, qi) ecdf(ti$V)(qi), d, q)
    FAV      = mapply(function(ti, qi) ecdf(ti$AV)(qi), d, q)

    # Determine di = FAV(q) - FA(q) - FV(q)
    di       = matrix(FAV - FA - FV, nrow=length(quantiles), ncol=length(d))

    # Observed test statistic
    stat     = tmax(di)

    # Permutation distribution of test statistic
    stati = numeric(nperm)
    for(i in 1:nperm)
        stati[i] = tmax(di %*% diag(sign(rnorm(d))))

    list(tmax=stat, tcrit=quantile(stati, 0.95), P=mean(stat <= stati))
} # rmi_perm

# Read in data from text file
#
# obs: observer number, cond: condition, rt: response time (Inf: omitted response)
#
# obs cond   rt  # This header should be included in the file
#   1    A  245  # response to A
#   1    V  281  # response to V
#   1   AV  212  # response to AV
#   1    V  Inf  # omitted response

rmi_read = function(fname)
{
    d = read.table(fname, header=TRUE)
    obs = split(d[, c('cond', 'rt')], d$obs)
    for(i in 1:length(obs))
        obs[[i]] = split(obs[[i]]$rt, obs[[i]]$cond)
    obs
} # rmi_read

# Example session with data in C:\RTDATA\go2004.dat

data = rmi_read('C:/RTDATA/go2004.dat')
rmi_perm(data, quantiles=c(.05, .10, .15, .20, .25, .30), nperm=10001)

File: external/Gondan-BRM-2009/rmiperm.m
function [p, tmax, tcrit] = rmiperm(fname, onlyobs, q, nperm)
% Permutation test of the race model inequality
%
% fname: filename to be read
%
%     obs cond    rt
%       1    A   234
%       1   AV   263
%       1    V   298
%       1    A   Inf
%
%     obs:  Observer number
%     cond: Condition (A, V, AV)
%     rt:   response time
%     Inf:  Omitted response
%
% onlyobs (optional): restrict analysis to these observers in the analysis
%     default []: use all observers
%
% q: quantiles of overall response time distribution at which RMI should be 
%    tested, default 0.05:0.05:0.30
%
% nperm: number of permutations used to determine tmax distribution
%     default 10001
%
% Tested with Matlab version 7.0.1
%

% defaults if parameters are omitted

if nargin < 1
    fname = 'c:\rtdata\go2004.dat' 
end

if nargin < 2
    onlyobs = [] 
end

if nargin < 3
    q = 0.05:0.05:0.30 
end

if nargin < 4
    nperm = 10001
end

% default return values

p = -1 ;
tcrit = -1 ;
tmax = 'error' ;

% read file

[fid, message] = fopen(fname, 'r') ;
if fid == -1
    disp(message)
    return
end

header = textscan(fid, '%s %s %s', 1) ;
if header{1}{1} ~= 'obs'
    disp('Error reading header.')
    return
end

if header{2}{1} ~= 'cond'
    disp('Error reading header.')
    return
end

if header{3}{1} ~= 'rt'
    disp('Error reading header.')
    return
end

data = textscan(fid, '%f %s %f', 'headerlines', 1) ;

fclose(fid) ;

% column 1: observer no
% column 2: condition (A, V, AV)
% column 3: response time

obs  = data{1} ;
cond = data{2} ;
rt   = data{3} ;

% Extract response times

if isempty(onlyobs)
    onlyobs = min(obs):max(obs) ;
end

d = [] ;

for i = onlyobs
    tA  = rt(obs == i & strcmp('A', cond)) ;
    tV  = rt(obs == i & strcmp('V', cond)) ;
    tAV = rt(obs == i & strcmp('AV', cond)) ;
    tAll = sort(rt(obs == i)) ;

    % Quantiles of overall RTs
    edges = tAll(round(q*length(tAll))) ;

    % Empirical CDFs
    maxrt = 1000 ;                  % adjust if necessary
    fA   = histc(tA, 1:maxrt) ;
    fV   = histc(tV, 1:maxrt) ;
    fAV  = histc(tAV, 1:maxrt) ;

    cdfA  = cumsum(fA)/length(tA) ;
    cdfV  = cumsum(fV)/length(tV) ;
    cdfAV = cumsum(fAV)/length(tAV) ;

    % Race Model Difference
    d = [d ; cdfAV(edges)' - cdfA(edges)' - cdfV(edges)'] ;
end

% Observed tmax statistic

tmax = max(mean(d) ./ std(d)) * sqrt(length(onlyobs)) ;

% Distribution of tmax

tmaxi = zeros(1, nperm) ;
for i = 1:nperm
    dp       = diag(sign(rand(1, length(onlyobs)) - 0.5)) * d ;
    tmaxi(i) = max(mean(dp) ./ std(dp))  * sqrt(length(onlyobs)) ;
end

% 95th quantile of simulated tmax

tmaxi = sort(tmaxi) ;
tcrit = tmaxi(round(0.95*nperm)) ;

% P value

p = sum(tmax < tmaxi)/nperm ;

if nargout < 3
    disp(sprintf('tmax=%.3f tcrit=%.3f P=%.3f', tmax, tcrit, p))
end

return
